### ISSUE TITLE
Regent: CUDA analysis behaves incorrectly when indexing into arrays.

### ISSUE THREAD

**Issue by ratoru** [StanfordLegion/legion#1124]
For the code below annotating the task with `__demand(__cuda)` does not raise an error. Yet, if I uncomment the `__demand(__openmp)` line, I get an error: “OpenMP code generation failed: found a loop-carried dependence.” According to @elliottslaughter this should always raise an error since the compiler doesn't know that the `rect2d` is shallow.

Changing `range_1d` to a `rect1d` and casting `i` as an `int` (as per #1121) exhibits the same behavior (= no error without `__demand(__openmp)`, but error with). The error given is: "OpenMP code generation failed: found an inadmissible statement." I believe this should give no error at all.
 
```
__demand(__cuda)
task example()
  var n = 10
  var range_1d = rect2d { int2d {0, 0}, int2d {n - 1, 0} }
  -- var range_1d = rect1d {0, n - 1}
  var arr : double[n]
 
  --__demand(__openmp)
  for i in range_1d do
    arr[i.x] = 1
    -- arr[int(i)] = 1
  end
end
``` 

When mixing arrays and regions in the same for loop, this leads to unmanageable errors. I can add an example for this, too, if needed.

**Comment by magnatelee** [comment]
The CUDA analysis not raising any error for the first example is actually an expected result, and to understand why, you need to understand a couple of things about the CUDA code generation. First, what the code generator does is to convert loops in a task into CUDA kernels. Parts that are not converted, including the loops that don't get mapped to kernels, run on the host side. Second, every Regent task is platform neutral, meaning there's no way you can do a GPU specific thing in a task, and the only way for a task to get a GPU allocation is via a region. This implies that any loop with no region access can't really run on GPUs, so the code generator will keep them intact for the host-side execution. That's why the checker concluded that the loop in your first example is "fine" (because it would run on the host side), but raised an error in the second example.

What you really need in both examples is a copy-out for the stack-allocated array `a` (which is not directly accessible from GPU kernels). Such a compiler transformation might be useful, but I haven't implemented it yet, because any loop that is worth running on GPUs will normally update a descent amount of data, which better be stored in a region. (and you would need much more than just the copy-out semantics at the end of every such loop for the best performance.)

I guess what's also missing here is a way of saying something like "all loops in a CUDA task must be mapped to kernels", and I'm open to a discussion on whether such a compiler pragma would be useful to add or not.

**Comment by elliottslaughter** [comment]
To be clear, does OpenMP support looping over rects? Because this code still seems to fail:

```
import "regent"

task example()
  var n = 5
  var range_1d = rect1d {0, n - 1}
  var arr : double[n] = [terralib.constant(`arrayof(double, 0, 1, 2, 3, 4))]
 
  __demand(__openmp)
  for i in range_1d do
    -- arr[i.x] = 1
    arr[int(i)] = 1
  end
end
```

**Comment by magnatelee** [comment]
Yes, it does. The thing that isn't supported is assignments to stack-allocated arrays. For example, this code works:
```
var sum = 0                                                                                                                                                                                                                                                                                                                                                                                                     
  __demand(__openmp)
for i in range_1d do
  sum += arr[int(i)]
end                                                                                                                                                                                                                                                                                                                                                                                                                           
```
I guess we could relax the checker to allow this only in OpenMP loops.

**Comment by elliottslaughter** [comment]
I guess there are two issues:

 1. Supporting arrays in OpenMP loops.
 2. The inaccuracy/unspecificity in the error message reporting (both in the line number and failing to indicate the array assignment is at fault) which makes it hard to figure out what's actually wrong.

I think we can work around (1) by just using a region, so I'm not too worried about it (but @ratoru let me know if you disagree). It would be nice to address (2) because at the moment there's a certain number of issues that need to be diagnosed by people who are familiar with the implementation because it's just not obvious what's going on right now.

**Comment by elliottslaughter** [comment]
Oh, I guess I should add:

 3. It would be nice to get a diagnostic that informs users about what loops are being assigned to the GPU. Even if we don't necessarily fix what was originally reported, it would be nice to at least know what loops are GPU-executable.
